src/datasources/redis/datasource.ts

```diff
@@ -158,18 +159,59 @@ export class PCPRedisDatasource {
         return label;
     }
 
+    async getTagKeys() {
+        const result = await this.pmSeriesSrv.getLabelsNames();
+        const keys = result.map(x => ({
+            type: 'string',
+            text: x
+        }));
+        return keys;
+    }
+
+    async getTagValues(option: any) {
+        const result = await this.pmSeriesSrv.getLabelValues([option.key]);
+        const values = result[option.key].map(x => ({ text: x }));
+        return values;
+    }
+
+    addAdHocFiltersToExpr(expr: string, filters: AdHocFilter[]) {
+        if (filters.length === 0) {
+            return expr;
+        }
+        const openingBracketIndex = expr.indexOf('{');
+        const hasMetadata = openingBracketIndex !== -1;
+        const filterCount = filters.length;
+        const adHocFilterStr = filters.reduce((previousValue, filter, index) => {
+            const modifiedOperator = filter.operator === DashboardVariableFilterOperator.Equals ? '==' : filter.operator;
+            const comma = index !== filterCount - 1 ? ',' : '';
+            const isNumericValue = !isNaN(parseFloat(filter.value)) && isFinite(+filter.value);
+            const formattedValue = isNumericValue ? filter.value : `"${filter.value}"`;
+            return `${previousValue}${filter.key}${modifiedOperator}${formattedValue}${comma}`;
+        }, '');
+        let newExpr;
+        if (hasMetadata) {
+            const appendPoint = openingBracketIndex + 1;
+            newExpr = `${expr.substring(0, appendPoint)}${adHocFilterStr},${expr.substring(appendPoint)}`;
+        } else {
+            newExpr = `${expr}{${adHocFilterStr}}`;
+        }
+        return newExpr;
+    }
+
     async query(query: Query) {
         const targets = this.buildQueryTargets(query);
         if (targets.length === 0)
             return { data: [] };
         if (!_.every(targets, ['format', targets[0].format]))
             throw new Error("Format must be the same for all queries of a panel.");
-
+        const datasourceName = this.name;
+        const variables = this.templateSrv.variables;
+        const adHocFilters = getAdHocFilters(datasourceName, variables);
         const exprs = targets.map(target => target.expr);
-        const series = await Promise.all(exprs.map(expr => this.pmSeriesSrv.query(expr)));
+        const exprsWithAdHocFiltersApplied = exprs.map(expr => this.addAdHocFiltersToExpr(expr, adHocFilters));
```

Comment:
again nitpicking here, you can chain the maps
```
const exprs = targets
    .map(target => target.expr)
    .map(expr => this.addAdHocFiltersToExpr(expr, adHocFilters));
```

Reply:
ah, then the `seriesByExpr` doesn't work anymore, because it matches the expr from the query editor... ignore the previous comment ;)